### PR TITLE
Usage of Gateway protected with Basic Auth

### PR DESCRIPTION
**Pull request by bartsmykla**
Closes https://github.com/openfaas-incubator/kafka-connector/issues/5

I created Request Builder abstraction so we don't have to put basic auth
everywhere where we want to do request. I have also changed whole codebase to
instead of http.NewRequest use requestBuilder.Build.

Right now at the beginning of function there is sent request for checking
Gateway Version.

Tests:
- request builder is fully tested
- getting gateway version is fully tested
I have created test_helpers with mocks of httpClient (which is being used
inside of GetGatewayVersion function instead of http.Client) and RequestBuilder
mock also.

I have tested changes on EKS cluster (2x t2.medium) by creating `kafka`
namespace and deploying there kafka with kafka-connector. I have added
basic-auth secrets (the same as inside of openfaas namespace) and then signed
into kafka console and created few messages.

Other:
I have added .DS_Store and .idea to .gitignore file

Signed-off-by: Bart Smykla (VMware) <bartek@smykla.com>

### CODE REVIEW

File: types/request_builder.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	Credentials RequestCredentials
+}
+
+type RequestBuilder interface {
+	Build(method string, url string, body io.Reader) (*http.Request, error)
+	BuildWithFuncPath(method string, funcPath string, body io.Reader) (*http.Request, error)
+	BuildWithURLBuilder(method string, buildURLFn func(string) string, body io.Reader) (*http.Request, error)
+}
+
+func (r *Request) Build(method string, url string, body io.Reader) (*http.Request, error) {
```
Comment (alexellis) [inline types/request_builder.go]: Not sure we need to repeat this line in all 3 functions

File: README.md [hunk 2]
```diff
@@ -19,6 +19,9 @@ When the connector hears a message on an advertised topic it will look that up i
 
 The cache or list of functions <-> topics is refreshed on a periodic basis.
 
+Kafka connector is working with Gateways protected using Basic Auth.
```
Comment (alexellis) [inline README.md]: I've seen `is working` in a few docs update recently. It's not the correct grammar for this kind of document. Perhaps write it using a simple phrase or no phrase at all?

I.e.

Basic Authentication on the gateway is supported.

File: README.md [hunk 3]
```diff
@@ -19,6 +19,9 @@ When the connector hears a message on an advertised topic it will look that up i
...
+If your Gateway is protected using Basic Auth want to deploy kafka connector inside other than openfaas namespace (where basic-auth secrets are also deployed). You need to remember that these secrets need to be deployed in your desired namespace also.
```
Comment (alexellis) [inline README.md]: Skip this paragraph and instruct them to install it to the correct namespace.

File: main.go [hunk 4]
```diff
@@ -79,7 +98,13 @@ func synchronizeLookups(ticker *time.Ticker,
 	}
 }
 
-func makeConsumer(client sarama.Client, brokers []string, config connectorConfig, topicMap *types.TopicMap) {
+func makeConsumer(
+	_ sarama.Client,
```
Comment (alexellis) [inline main.go]: why did you change this to `_`?

File: types/credentials.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+	val, err := readFile("/var/secrets/basic-auth-user")
+
+	if err == nil {
+		credentials.Username = val
+	} else {
+		log.Printf("Unable to read username: %s", err)
+	}
+
+	passwordVal, passErr := readFile("/var/secrets/basic-auth-password")
```
Comment (alexellis) [inline types/credentials.go]: You can't hard-code the path here. This needs to read a secrets override path like the gateway does. Please cross-reference how the gateway approaches this.

**Comment by bartsmykla** [comment]
@alexellis: ETA 1-2h

**Comment by bartsmykla** [comment]
@alexellis I have removed abstraction, and all modifications not related with tasks. Can you review it again?

File: types/credentials.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+	if len(secretPath) == 0 {
+		secretPath = "/var/secrets"
+	}
+
+	return filepath.Join(secretPath, secretName)
+}
+
+func GetCredentials() *Credentials {
+	credentials := Credentials{}
```
Comment (alexellis) [inline types/credentials.go]: Are we not able to vendor this from the openfaas/faas-provider?

**Comment by alexellis** [comment]
I'm a bit confused at the list of vendored files.. can you check these are all required?

i.e. `vendor/github.com/rcrowley/go-metrics/debug.go` and the webhookstash sample function from `openfaas/faas`?

**Comment by alexellis** [comment]
Can we not vendor the Credentials type from here rather than having to copy/paste or re-implement it?

https://github.com/openfaas/faas-provider/blob/master/auth/credentials.go

**Comment by alexellis** [comment]
@bartsmykla please can you work on the feedback and do a rebase since that's been blocking any potential merge since by comment a month ago. 

Let me know if you have any questions.

Alex
